### PR TITLE
Ingest pipeline

### PR DESCRIPTION
**Pull request by joverlee521**
The ingest pipeline is a Snakemake pipeline split into 4 modules:
1. `fetch_sequences.smk` - fetch and concat all sequence data into a single `sequences.ndjson` file.
      - this currently only fetches from GenBank, but we can easily add other sources. 
      - other sources can be added as new rules or just by adding files as `ingest/data/{source}.ndjson`
2. `transform.smk` - all data transformations 
      - currently only one `transform` rule with a shell pipeline, but as @tsibley pointed out in [Slack](https://bedfordlab.slack.com/archives/C03G8HQEV18/p1653590914450019?thread_ts=1653529586.342209&cid=C03G8HQEV18), we can split this into multiple rules using Snakemake's [piped outputs](https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#piped-output).
3. `upload.smk` - upload files in `ingest/data/` to remote destinations
      - currently only supports uploads to AWS S3 
      - uses predefined wildcards for local and remote files so each job is a single file upload that can run as soon as the local file has been created
 4. `slack_notifications.smk` - Slack notification rules intended for Nextstrain team's internal use only 

---
The configs for the pipeline is split into 2 config YAML files:
1. `ingest/config/config.yaml` - contains the required pipeline configs 
      - I pulled out all of the transform parameters into this config file so that they can be updated without editing the transform scripts
2. `ingest/config/optional.yaml` - contains optional configs for upload rules and Slack notifications 

--- 
The pipeline is automated via GitHub Actions workflow that launches the ingest job on AWS Batch. It is scheduled to run daily at 16:00 UTC (09:00 PT), but we can increase the frequency if desired. It can also be triggered manually via the Actions UI or a repository dispatch. 

See test run:
- [GH action](https://github.com/nextstrain/monkeypox/actions/runs/2431858083) 
- AWS Batch Job id `bf0de508-2188-4f7e-b318-aa982741bc81`
- notifications in Slack channel `#monkeypox-updates`
- files at `s3://nextstrain-data/files/workflows/monkeypox/`

---
Future TODOs:
- [ ] canonicalize strain names
- [ ] deduplication based on strain names (?)
- [ ] adopt ncov-ingest's [geolocation rules](https://github.com/nextstrain/ncov-ingest/blob/master/source-data/gisaid_geoLocationRules.tsv) and [location hierarchy](https://github.com/nextstrain/ncov-ingest/blob/master/source-data/location_hierarchy.tsv) (?)
- [ ] thread notifications (?)
- [x] include full authors list for acknowledgements table (requested by @chaoran-chen) - added in 5985b98
- [ ] run NextClade and include output in metadata TSV (requested by @chaoran-chen)

### CODE REVIEW

**Comment by emmahodcroft** [comment]
I think this is looking great @joverlee521 , thank you for working on this!

IMO, the most important things to get finalized before we merge would be:
[ ] adopt ncov-ingest's geolocation rules _(just makes everything else much easier & I think should be quite portable?)_
[ ] run NextClade and include output in metadata TSV

We should decide on whether we want to switch to using an ID as our `strain` in the Nextstrain workflow now. This would allow us to put off worrying about strain names until later (as they they just get added back at the viz end and don't need to be unique for the pipeline). This is mostly straightforward except...

We have one sequence which isn't in Genbank (`mpx-37-Finland_consensus`). If we want to include it:

[ ] Add extra ingest file with the Finnish sequence & add a `LPS` id
[ ] Figure out how to set which ID is used for fasta files & for the pipeline (ex: if the Finnish seq later gets GB, do we switch to using GB?)
 
 And these we can work on more after merge:
 [ ] canonicalize strain names
 [ ] deduplication based on strain names (?)

File: ingest/bin/transform-authors
Comment (victorlin) [inline ingest/bin/transform-authors:1]: *(minor suggestion)*

use `ingest/transform:` in commit title for consistency

**Review by victorlin — COMMENTED** [review]
Reviewed through the transform scripts, just a few minor things but otherwise looks good! Will look at the rest after lunch.

File: ingest/README.md
Comment (corneliusroemer) [inline ingest/README.md:13]: This doesn't seem to run the ingest pipeline but the general Snakemake workflow instead, when my current working directory is at the top-level. Maybe we should add a sentence that one needs to be in the ingest folder?
Something like:
```sh
nextstrain build ingest/
```

**Comment by rneher** [comment]
I think that can be merged as soons as the little things are sorted. the nextclade info can be added in a second PR if need be after @corneliusroemer has updated the reference sequence.

File: ingest/Snakefile
Comment (corneliusroemer) [inline ingest/Snakefile:50]: Why don't we simply put the targets here? The function above feels a bit overkill for the current simple ingest flow? Do we have anything in mind other than the default targets?

File: ingest/Snakefile
Comment (victorlin) [inline ingest/Snakefile:32]: *(minor suggestion)*

```suggestion
                    [f"data/upload/{target}/{{file_to_upload}}-to-{{remote_file_name}}.done"],
```

File: ingest/Snakefile
Comment (victorlin) [inline ingest/Snakefile:36]: This is my first time looking at Snakemake's built-in [`expand()`](https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#the-expand-function). For this specific usage, is it just the same as this?

```py
[f"data/upload/{target}/{file_to_upload}-to-{remote_file_name}.done"
    for file_to_upload, remote_file_name in
    zip(files_to_upload, remote_file_names)]
```

File: ingest/workflow/snakemake_rules/transform.smk
Comment (victorlin) [inline ingest/workflow/snakemake_rules/transform.smk:67]: Wow, this is so much easier to follow than [ncov-ingest/Snakefile](https://github.com/nextstrain/ncov-ingest/blob/1eb377df20670a896a3942b1a8eed78df15ccd18/Snakefile)!

**Review by victorlin — APPROVED** [review]

File: ingest/bin/transform-date-fields
Comment (joverlee521) [reply ingest/bin/transform-date-fields:24]: Ah, yeah I copied this function from ncov-ingest without modifying the doctests. I added the `WARNING` messages for this specific script. I would add CRAM tests for these if we officially add them to augur.

File: ingest/bin/transform-genbank-location
Comment (joverlee521) [reply ingest/bin/transform-genbank-location:26]: This is the general pattern seen in SARS-CoV-2 data that I'm hoping people will also follow in monkeypox.
The [GenBank country qualifier docs](https://www.ncbi.nlm.nih.gov/genbank/collab/country/) show this value format: `"<country_value>[:<region>][, <locality>]"`

File: ingest/Snakefile
Comment (joverlee521) [reply ingest/Snakefile:36]: It's basically doing the list comprehension that you have, but it also adds `file_to_upload` and `remote_file_name` as Snakemake wildcards. I'm then using these wildcards in [upload.smk](https://github.com/nextstrain/monkeypox/blob/ingest/ingest/workflow/snakemake_rules/upload.smk) to create a job for each file upload. 

This is also the part of the pipeline that I'm most unsure about. It works, but I'm not sure if this is a good Snakemake practice...

File: ingest/workflow/snakemake_rules/transform.smk
Comment (joverlee521) [reply ingest/workflow/snakemake_rules/transform.smk:67]: Yay, that's good to hear! The transform rule was my attempt to simplify [ncov-ingest/bin/transform-genbank](https://github.com/nextstrain/ncov-ingest/blob/1eb377df20670a896a3942b1a8eed78df15ccd18/bin/transform-genbank).

File: ingest/README.md
Comment (joverlee521) [reply ingest/README.md]: Ah yup, I was assuming if reading the `ingest/README` you were already in the ingest directory. 
I can add more language to clarify this!

File: ingest/Snakefile
Comment (joverlee521) [reply ingest/Snakefile:50]: Hmm, this was my way of making the upload/Slack notification targets optional but also easily included through parsing the config. Is there a better way to do optional targets?

File: ingest/README.md
Comment (joverlee521) [reply ingest/README.md:22]: Yup, I'll add more language to clarify.

File: ingest/bin/fetch-from-genbank
Comment (joverlee521) [reply ingest/bin/fetch-from-genbank]: Thanks for catching that! Fixed up in e15a08d972271314fa9ace37230b0c79368dd394

File: ingest/bin/fasta-to-ndjson
Comment (joverlee521) [reply ingest/bin/fasta-to-ndjson]: Yup, fixed up in cb5daff4ee873478038da90119172a27083b47e6.

Good to know this will be available augur!

File: ingest/bin/transform-strain-names
Comment (joverlee521) [reply ingest/bin/transform-strain-names]: Thank you! Fixed up in 7c3106b3f33a2373693c666accde1544eb22b7cf

File: ingest/bin/transform-string-fields
Comment (joverlee521) [reply ingest/bin/transform-string-fields]: Thanks! Fixed up in dbc5f8d188c5ee2883b3eb6b6d526051dad18249.

(Also added a check that they were provided as args!)

File: ingest/bin/transform-string-fields
Comment (joverlee521) [reply ingest/bin/transform-string-fields]: Sounds reasonable to me. Fixed up in dbc5f8d188c5ee2883b3eb6b6d526051dad18249

File: ingest/bin/transform-authors
Comment (joverlee521) [reply ingest/bin/transform-authors]: Sounds reasonable to me. Fixed up in 43483b385f8978fe2fc06824517ee9692ab25f7d

File: ingest/bin/transform-date-fields
Comment (joverlee521) [reply ingest/bin/transform-date-fields:10]: Fixed up in a7428dc199d127f9d3f8e3598b7c56935993a248. (I need a linter 😄)

File: ingest/bin/transform-authors
Comment (joverlee521) [reply ingest/bin/transform-authors:1]: ❤️ consistency. Reworded in ff61804c815c5cddc19360e41eeb4e60dfffe21f

File: ingest/bin/merge-user-metadata
Comment (joverlee521) [reply ingest/bin/merge-user-metadata]: Fixed up in 622bd1498f11467c4ae01590a41587e6a661eb40. Also reworded commit message for consistency.

File: ingest/bin/merge-user-metadata
Comment (joverlee521) [reply ingest/bin/merge-user-metadata]: Ah thank you, I learned about `partition` a while back and keep forgetting to use it.
Fixed up in 622bd1498f11467c4ae01590a41587e6a661eb40

File: ingest/Snakefile
Comment (joverlee521) [reply ingest/Snakefile]: Fixed up in 50d4662e796bcfbba33f5d1f85c3a0813ed06804

**Review by victorlin — APPROVED** [review]
Thanks for addressing everything! Just a couple more follow-ups but overall 👍 this can be merged to use sooner than later.

File: ingest/bin/transform-genbank-location
Comment (joverlee521) [reply ingest/bin/transform-genbank-location]: Yup, added a comment in 2f594549be7de9cc43343bb2899e9067be11b63c

**Comment by joverlee521** [comment]
I've created [multiple issues](https://github.com/nextstrain/monkeypox/issues?q=is%3Aissue+is%3Aopen+ingest+created%3A2022-06-07+author%3Ajoverlee521) to track the outstanding TODOs from this PR.

**Pull request merged**
